server/src/grpc/storage_service/query.rs

```diff
@@ -43,10 +51,66 @@ fn empty_ok_resp() -> QueryResponse {
     }
 }
 
+async fn maybe_forward_query<Q: QueryExecutor + 'static>(
+    ctx: &HandlerContext<'_, Q>,
+    req: &QueryRequest,
+) -> Option<Result<QueryResponse>> {
+    let forwarder = ctx.forwarder.as_ref()?;
+
+    if req.metrics.len() != 1 {
+        warn!(
+            "Unable to forward query without exactly one metric, req:{:?}",
+            req
+        );
+
+        return None;
+    }
+
+    let forward_req = ForwardRequest {
+        schema: ctx.schema.clone(),
+        metric: req.metrics[0].clone(),
+        req: req.clone().into_request(),
+    };
+    let do_query = |mut client: StorageServiceClient<Channel>,
+                    request: tonic::Request<QueryRequest>,
+                    _: &Endpoint| {
+        let query = async move {
+            client
+                .query(request)
+                .await
+                .map(|resp| resp.into_inner())
+                .map_err(|e| Box::new(e) as _)
+                .context(ErrWithCause {
+                    code: StatusCode::INTERNAL_SERVER_ERROR,
+                    msg: "Forwarded query failed".to_string(),
+                })
+        }
+        .boxed();
+
+        Box::new(query) as _
+    };
+
+    match forwarder.forward(forward_req, do_query).await {
+        Ok(forward_res) => match forward_res {
+            ForwardResult::Forwarded(v) => Some(v),
+            ForwardResult::Original => None,
+        },
+        Err(e) => {
```

Comment:
why not return error directly when this fail?

Reply:
For error caused by `Forward` itself, I guess we should give it a try to the local service.